### ISSUE TITLE
Make FormBuilder Projects More Portable

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1142]
When I create a FormBuilder project (Domino On Disk Project) on my local machine, I see multiple absolute paths in the `.ondiskproj` file:

```
<project>
  ...
  <mavenBuild>
    ...
    <option mavenBuildPath="/Users/-user-/Documents/Flex/temp_workspace/TestPaths"/>
    ...
    <option settingsFilePath="/Users/-user-/Downloads/MoonshineSDKs/Domino/settings.xml"/>
    ...
  </mavenBuild>
  ...
  <domino>
    <option localDatabase="/Users/-user-/Documents/Flex/temp_workspace/TestPaths/nsfs/nsf-moonshine/target/nsf-moonshine-domino-1.0.0.nsf"/>
    ...
  </domino>
  ...
</project>
```

If I want to put these projects on GitHub, other users won't be able to use the project unless they fix the paths.  Since the paths are completely invalid, the file choosers will open to the last opened file, so the user will not get any help finding the files.

I found I could set `mavenBuildPath` to `.`, and it worked.  However, `localDatabase` did not support a relative path, even if I used `./`.

I'm not sure how to handle `settingsFilePath`, since this is not a relative path.  Maybe we need a special variable for this, but if it gets replaced on the individual machines then it would be an absolute path again.

UPDATE:  Branch:  features/issue_1142_path_setup
macOS: https://static.moonshine-ide.com/downloads/releases/dev/macos/branches/features/issue_1142_path_setup
Windows: https://static.moonshine-ide.com/downloads/releases/dev/windows/branches/features/issue_1142_path_setup

**Comment by rat-moonshine** [comment]
For a note, behind my intention to use absolute paths in settings/different-sections to support paths from different drives in Windows system. It's possible (maybe rather non-standard) that a project configured source from different drives from a Windows system. In that case I found, the relative path generation failed to properly address the paths. Maybe, I need to look into this again as those memory almost faded now.

Moonshine settings mostly relying upon absolute paths currently. If we able to address this issue with a generalise method, then perhaps we should update to other sections, as well. I shall check this shortly.

**Comment by JoelProminic** [comment]
So to clarify, if we have a project with references like this:

    C:/workspace/MainProject
    D:/workspace/SubProject
    
In this case, we can't have a relative path because of the different drives.

External project dependencies are problematic for relative paths in other cases.  I we have a relative path like `../../../another-workspace/SubProject`, this is probably not a useful relative path.  `../Subproject` is a common case I would like to support, though.

However, this is not directly relevant to this issue.  The file references in this case are subdirectories of the main project, except for `settingsFilePath`.  I don't want a relative path for `settingsFilePath` because something like `../../../Downloads/MoonshineSDKs/Domino/settings.xml` would not be consistent across all macOS environments, and it would be wrong for Windows.

**Comment by JoelProminic** [comment]
To simplify this for now:
- Store a relative path for any files inside the project directory
- Store an absolute path for files outside the project directory
- Support external paths (i.e. ../Subproject), and do not overwrite the values unless the user sets a new path

**Comment by JoelProminic** [comment]
Another annoyance for the .proj file is the sorting:
![image](https://user-images.githubusercontent.com/24532731/204909596-d0cced08-f50b-47d8-b547-2dc5f55686c1.png)

You can see differences here where `linkReport`, `commandLine`, `settingsFilePath`, and `targetDatabase` appear in a different order in the updated .as3proj file.  This makes it harder to tell what has actually changed when the file is updated in Git or SVN.  This is a more generic problem, so we can move it to a separate issue if necessarly.

**Comment by JoelProminic** [comment]
Note that currently if we use `.` for the `mavenBuildPath`, then a local build actually runs in the last opened directory from a file chooser. For example, I set `mavenSettingsPath` in a project and then ran Project > Build with Apache Maven, and I got this error as it ran in the wrong directory:

    : Maven build directory: /Users/joelanderson/Downloads/MoonshineSDKs/Domino
    : Command: /Users/joelanderson/Downloads/MoonshineSDKs/Maven/apache-maven-3.6.0/bin/mvn -settings "/Users/joelanderson/Downloads/MoonshineSDKs/Domino/settings.xml" clean install
    : [INFO] Scanning for projects...
    : [INFO] ------------------------------------------------------------------------
    : [INFO] BUILD FAILURE
    : [INFO] ------------------------------------------------------------------------
    : [INFO] Total time:  0.099 s
    : [INFO] Finished at: 2022-11-30T16:13:50-06:00
    : [INFO] ------------------------------------------------------------------------
    : [ERROR] The goal you specified requires a project to execute but there is no POM in this directory (/Users/joelanderson/Downloads/MoonshineSDKs/Domino). Please verify you invoked Maven from the correct directory. -> [Help 1]

We'll  need to handle this properly with the updated solution.

**Comment by rat-moonshine** [comment]
Currently testing logic populates the paths in an expected manner, i.e.

```
<mavenBuild>
    <option settingsFilePath="src/main/pom.xml"/>
    <option mavenBuildPath="."/>
```
The system is still in testing and it requires to test in different platforms/cases. Once written in the manner they also needs to be retrieved in a full-path way so Moonshine can use the paths Whereis necessary, i.e. in Build with Apache Maven case mentioned [earlier](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1142#issuecomment-1332803576).

**Comment by rat-moonshine** [comment]
Project Path | Source Path | Output Path
---------------- | --------------- | -------------
/Users/$user/Desktop/WorkspaceA/ProjectA | /Users/$user/Desktop/WorkspaceA/ProjectA | `.`
/Users/$user/Desktop/WorkspaceA/ProjectA | /Users/$user/Desktop/WorkspaceA/ProjectA/subFolder/abcd.txt | `subFolder/abcd.txt`
/Users/$user/Desktop/WorkspaceA/ProjectA | /Users/$user/Desktop/WorkspaceA/ProjectB/efgh.txt | `../ProjectB/efgh.txt`
/Users/$user/Desktop/WorkspaceA/ProjectA | /Users/$user/Desktop/WorkspaceB/ProjectC/ijkl.txt | `/Users/$user/Desktop/WorkspaceB/ProjectC/ijkl.txt`

**Comment by JoelProminic** [comment]
I tested some actions that use the settings:

Action | Settings | Result (macOS)
-- | -- | --
Build with Apache Maven | mavenBuildPath, settingsFilePath | Works
Deploy Database on Vagrant | localDatabase | Works
Clean | mavenBuildPath | Works



Update Tests:

Test | Expected | Actual (macOS)
-- | -- | --
New FormBuilder projects | Use Relative paths | Confirmed
Set "grandparent" relative path (`../../`).  Open a project | Do not update paths | Confirmed
Set "grandparent" relative path (`../../`).  Open and save the settings without changing any paths  | Do not update paths | BUG:  Paths updated from `../../` to `../`
Set a new path under the project directory | Save relative path | Confirmed
Set a new path under the project's parent directory | Save relative path | Confirmed
Set a new path undert the project's grandparent directory | Save absolute path | Confirmed

**Comment by rat-moonshine** [comment]
![image](https://user-images.githubusercontent.com/24370870/205555635-c8a8b836-1a9d-46cd-8d43-0f7cab245763.png)

The above problem should update now, but a settings save-logic will update the value with absolute correct path, since the current system supports only one `../` relative generation.

**Comment by JoelProminic** [comment]
I reran the tests with the updates, and they seem to work well so far.  I didn't have any trouble with `%NSFODP_SETTINGS_FILE%` in my tests.

Action | Settings | Result (macOS)
-- | -- | --
Build with Apache Maven | mavenBuildPath, settingsFilePath | Confirmed
Deploy Database on Vagrant | localDatabase | 
Clean | mavenBuildPath | 


Test | Expected | Actual (macOS)
-- | -- | --
New FormBuilder projects | Use Relative paths | Confirmed
Set relative path (`../`).  Open a project | Do not update paths | Confirmed
Set relative path (`../`).  Open and save the settings without changing any paths  | Do not update paths | Confirmed
Set "grandparent" relative path (`../../`).  Open a project | Do not update paths | Confirmed
Set "grandparent" relative path (`../../`).  Open and save the settings without changing any paths  | Update the paths to absolute paths | Confirmed
Set a "grandparent" relative path that does not exist.  Open and save the settings without changing any paths | Save the absolute path anyway | Cofirmed
Set a new path under the project directory | Save relative path | Confirmed
Set a new path under the project's parent directory | Save relative path | Confirmed
Set a new path undert the project's grandparent directory | Save absolute path | Confirmed

Note that the current design has limited support for `../..` paths. This doesn't really matter for FormBuilder projects, but there are cases in the Moonshine-IDE project that require two or more levels of `..` for a relative path.

**Comment by JoelProminic** [comment]
The changes could affect other Maven projects, so we should do some tests.

Otherwise, this is ready for a merge.

**Comment by rat-moonshine** [comment]
This issue is now merged into `master`.

**Comment by JoelProminic** [comment]
I reproduced the error on macOS 12.6.1 (Monterey)

    : [INFO] Compiling 1 source file to /Users/-user-/Documents/TestJavaMaven_12_15/target/classes
    : [INFO] -------------------------------------------------------------
    : [ERROR] COMPILATION ERROR : 
    : [INFO] -------------------------------------------------------------
    : [ERROR] Source option 6 is no longer supported. Use 7 or later.
    : [ERROR] Target option 6 is no longer supported. Use 7 or later.
    : [INFO] 2 errors 

I was able to resolve this by upgrading the plugins in pom.xml (which increases the default Java version to 1.7)

Plugin | Old Version | New Version
-- | -- | --
maven-compiler-plugin |  | 3.10.1
maven-jar-plugin | 3.2.0 | 3.3.0

```
<?xml version="1.0"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
  xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>

  <groupId>local</groupId>
  <artifactId>TestJavaMaven_1215</artifactId>
  <name>TestJavaMaven_1215</name>
  <version>0.1.0-SNAPSHOT</version>
   <build>
        <sourceDirectory>src/main/java</sourceDirectory>
        <plugins>
            <plugin>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.10.1</version>
            </plugin>
            <plugin>
                  <groupId>org.apache.maven.plugins</groupId>
                  <artifactId>maven-jar-plugin</artifactId>
                  <version>3.3.0</version>
                  <configuration>
                    <archive>
                      <manifest>
                        <mainClass>TestJavaMaven_1215</mainClass>
                      </manifest>
                    </archive>
                  </configuration>
                </plugin>
        </plugins>
   </build>
</project>
```

Project | Build & Run Test (macOS Monterey x64)
-- | -- 
Java Maven | Requires updates
Royale Browser, Jewel (Build with Apache Maven) | No default build actions or pom.xml.  I though this was supported.
Domino On Disk Project | Blocked by missing update site
Domino Visual Editor |  Blocked by missing update site
Settings > Domino and Notes Client > UpdateSite > Generate | Not enabled on macOS Monterey

**Comment by JoelProminic** [comment]
This is still waiting on Maven testing.

**Comment by JoelProminic** [comment]
Revisiting this for [Super.Human.Portal](https://github.com/Moonshine-IDE/Super.Human.Portal), I see that the update is not working for `localDatabase`:

    <option localDatabase="./nsfs/nsf-moonshine/target/nsf-moonshine-domino-1.0.0.nsf"/>
    
![image](https://user-images.githubusercontent.com/24532731/212437575-8a6b4994-1615-47a3-bcc6-2ed6ade95e3c.png)

I confirmed the path does exist relative to the project directory.

I also see additional absolute paths for [jsOutput](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/8d51ceb7b7ab088832dce5a5422e2dc01b31f5c4/Super.Human.Portal.veditorproj#L16) and [customSDK](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/8d51ceb7b7ab088832dce5a5422e2dc01b31f5c4/Super.Human.Portal.veditorproj#L44) - I removed these for now.

`mavenBuildPath` and `settingsFilePath` were set properly without input from me.

**Comment by rat-moonshine** [comment]
> <option localDatabase="./nsfs/nsf-moonshine/target/nsf-moonshine-domino-1.0.0.nsf"/>

1. How this saved in project configuration file?
2. Is this `./` path auto-generated and shows as it is in project's settings?
3. In Super.Human.Portal, if you try to check the `.nsf` path off-Moonshine what is the full file path? Does the path adds any foreign element during running inside Super.Human.Portal?

**Comment by rat-moonshine** [comment]
`./` relative syntax now handles. i.e. 

`<option localDatabase="./nsfs/nsf-moonshine/target/nsf-moonshine-domino-1.0.0.nsf"/>`

**Comment by JoelProminic** [comment]
I confirmed that `localDatabase` is working as expected for Domino Visual Editor

I see that there are other absolute paths in the `.visualeditorproj` file, but they did not cause any errors and could be safely removed.  This is already outside the scope of the original issue (FormBuilder), so these properties can be ignored.
```
  <jsOutput>
    <option path="/Users/joelanderson/Documents/Flex/temp_workspace/TestDominoVisualEditor_0117"/>
  </jsOutput>
  ...

    <option customSDK="/Users/joelanderson/Downloads/MoonshineSDKs/Flex_SDK/Flex_4.16.1_AIR_32.0"/>
```


I see this was previously waiting for some testing on other Maven projects

**Comment by JoelProminic** [comment]
For Java Maven projects, I see that `mavenBuild/mavenBuildPath` is not populated by default on project creation.

```
<?xml version="1.0" encoding="utf-8"?>
<project>
  <mavenBuild>
      <option commandLine="clean package"/>
    </mavenBuild>
  <classpaths>
    <class path="src/main/java"/>
  </classpaths>
  <build>
    <option mainclass="TestJavaMaven_0123"/>
  </build>
  <options>
    <option projectType="java-maven"/>
  </options>
</project>

```

If I save the settings, it does get populated with a relative path:

```
  <mavenBuild>
    <option dominoNotesProgram=""/>
    <option mavenBuildPath="."/>
    <option settingsFilePath=""/>
    <option dominoNotesPlatform=""/>
```

I also confirme that I could change the classpath to an external project in the parent directory, and it was saved as a relative path.  (Note that I had to manually [this fix](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1167#issuecomment-1353804088) to let the project build.
